QuantumToolbox.jl/time_evolution/Quantum_Fisher_Information_with_automatic_differentiation.qmd

```diff
@@ -0,0 +1,157 @@
+    c_ops = [sqrt(γ)*a]
+    ψ0 = fock(N, 0) # initial state
+
+    tlist = range(0, 2000, 100)
+
+    sol = mesolve(H, ψ0, tlist, c_ops; params = p, progress_bar = Val(false), saveat = [tlist[end]])
+    return sol.states[end].data
+end
+```
+
+```{julia}
+function final_state(p, t)
+    G, K, γ = 0.002, 0.001, 0.01
+
+    N = 20 # cutoff of the Hilbert space dimension
+    a = destroy(N) # annihilation operator
+
+    coef(p,t) = - p[1]
+    H = QobjEvo(a' * a , coef) + K * a' * a' * a * a - G * (a' * a' + a * a)
+    c_ops = [sqrt(γ)*a]
+    ψ0 = fock(N, 0) # initial state
+
+    tlist = range(0, 2000, 100)
+
+    sol = mesolve(H, ψ0, tlist, c_ops; params = p, progress_bar = Val(false), saveat = [t])
+    return sol.states[end].data
+end
+```
+
+## Quantum Fisher Information Calculation
+
+The QFI is computed using the symmetric logarithmic derivative (SLD). For a density matrix $\rho(\theta)$ parametrized by $\theta$:
+
+$$F_Q = \text{Tr}[\partial_\theta \rho \cdot L]$$
+
+where $L$ is the SLD satisfying: $\partial_\theta \rho = \frac{1}{2}(\rho L + L \rho)$
+
+```{julia}
+function compute_fisher_information(ρ, dρ)
+    # Add small regularization to avoid numerical issues with zero eigenvalues
+    reg = 1e-12 * I
+    ρ_reg = ρ + reg
+
+    # Solve for the symmetric logarithmic derivative L
+    # dρ = (1/2)(ρL + Lρ)
+    # This is a Sylvester equation: ρL + Lρ = 2*dρ
+    L = sylvester(ρ_reg, ρ_reg, -2*dρ)
+
+    # Fisher information F = Tr(dρ * L)
+    F = real(tr(dρ * L))
+
+    return F
+end
+```
+
+## Automatic Differentiation Setup
+
+We use finite differences through DifferentiationInterface.jl to compute the derivative of the quantum state with respect to the parameter. This is a key step that enables efficient QFI computation without manual derivative calculations.
+
+```{julia}
+# Test the system
+final_state([0], 100)
+
+# Define state function for automatic differentiation
+state(p) = final_state(p, 2000)
+
+# Compute both the state and its derivative
+ρ, dρ = DifferentiationInterface.value_and_jacobian(state, AutoFiniteDiff(), [0.0])
+
+# Reshape the derivative back to matrix form
+dρ = QuantumToolbox.vec2mat(vec(dρ))
+
+# Compute QFI at final time
+qfi_final = compute_fisher_information(ρ, dρ)
+println("QFI at final time: ", qfi_final)
+```
+
+## Time Evolution of Quantum Fisher Information
+
+Now we compute how the QFI evolves over time to understand the optimal measurement time for parameter estimation:
+
+```{julia}
+ts = range(0, 2000, 100)
+
+QFI_t = map(ts) do t
+    state(p) = final_state(p, t)
+    ρ, dρ = DifferentiationInterface.value_and_jacobian(state, AutoFiniteDiff(), [0.0])
+    dρ = QuantumToolbox.vec2mat(vec(dρ))
+    compute_fisher_information(ρ, dρ)
+end
+
+println("QFI computed for ", length(ts), " time points")
+```
+
+## Visualization
+
+Plot the time evolution of the Quantum Fisher Information:
+
+```{julia}
+fig = Figure(size=(900, 350))
```

Comment:
Could you remove the size here? We should have set a default size in Quarto already.